mods/doors/init.lua

```diff
@@ -1,435 +1,444 @@
+			end,
+			state = function(self)
+				local name = minetest.get_node(pos).name
+				if string.sub(name, -5) == "_open" then
+					return true
+				else
+					return false
+				end
+			end
+		}
+	else
+		return nil
 	end
-	if not def.selection_box_top then
-		def.selection_box_top = box
+end
+
+-- this hidden node is placed on top of the bottom, and prevents
+-- nodes from being placed in the top half of the door.
+minetest.register_node("doors:hidden", {
+	description = "Hidden Door Segment",
+	drawtype = "airlike",
+	paramtype = "light",
+	sunlight_propagates = true,
+	walkable = false,
+	pointable = false,
+	diggable = false,
+	buildable_to = false,
+	floodable = false,
+	drop = "",
+	groups = { not_in_creative_inventory = 1 },
+	on_blast = function() end
+})
+
+-- table used to aid door opening/closing
+local transform = {
+	{
+		{ v = "_a", param2 = 3 },
+		{ v = "_a", param2 = 0 },
+		{ v = "_a", param2 = 1 },
+		{ v = "_a", param2 = 2 },
+	},
+	{
+		{ v = "_b", param2 = 1 },
+		{ v = "_b", param2 = 2 },
+		{ v = "_b", param2 = 3 },
+		{ v = "_b", param2 = 0 },
+	},
+	{
+		{ v = "_b", param2 = 1 },
+		{ v = "_b", param2 = 2 },
+		{ v = "_b", param2 = 3 },
+		{ v = "_b", param2 = 0 },
+	},
+	{
+		{ v = "_a", param2 = 3 },
+		{ v = "_a", param2 = 0 },
+		{ v = "_a", param2 = 1 },
+		{ v = "_a", param2 = 2 },
+	},
+}
+
+function _doors.door_toggle(pos, clicker)
+	local meta = minetest.get_meta(pos)
+	local state = meta:get_int("state")
+	local def = minetest.registered_nodes[minetest.get_node(pos).name]
+	local name = def.door.basename
+
+	if clicker ~= nil then
+		local owner = meta:get_string("doors_owner")
+		if owner ~= "" then
+			if clicker:get_player_name() ~= owner then
+				return false
+			end
+		end
 	end
 
-	if not def.sound_close_door then
-		def.sound_close_door = "doors_door_close"
+	local old = state
+	-- until Lua-5.2 we have no bitwise operators :(
+	if state % 2 == 1 then
+		state = state - 1
+	else
+		state = state + 1
 	end
-	if not def.sound_open_door then
-		def.sound_open_door = "doors_door_open"
+
+	local dir = minetest.get_node(pos).param2
+	if state % 2 == 0 then
+		minetest.sound_play(def.door.sounds[1], {pos = pos, gain = 0.3, max_hear_distance = 10})
+	else
+		minetest.sound_play(def.door.sounds[2], {pos = pos, gain = 0.3, max_hear_distance = 10})
 	end
-	
-	
-	minetest.register_craftitem(name, {
+
+	minetest.swap_node(pos, {
+		name = "doors:" .. name .. transform[state + 1][dir+1].v,
+		param2 = transform[state + 1][dir+1].param2
+	})
+	meta:set_int("state", state)
+
+	return true
+end
+
+function doors.register(name, def)
+	-- replace old doors of this type automatically
+	minetest.register_abm({
+		nodenames = {"doors:"..name.."_b_1", "doors:"..name.."_b_2"},
+		interval = 5.0,
+		chance = 1,
+		catch_up = true,
+		action = function(pos, node, active_object_count, active_object_count_wider)
+			local l = tonumber(string.sub(node.name, string.len(node.name)))
```

Comment:
You can use -1 instead of `string.len`. Also, `node.name:sub(...)`, please. If it's really performance critical, localize.